mssql_base.py

```diff
@@ -1,91 +1,85 @@
-                        default=False, order=6, advanced=True)
-
-    def __init__(self):
-        super().__init__()
-        self.cnxn = None
-        self.cursor = None
-        self.isConnecting = False
-        # maintains a LUT index by table containing a list of columns for it
-        self._table_colums = {}
-
-    def configure(self, context):
-        super().configure(context)
-        self.connect()
-
-    def connect(self):
-        self.isConnecting = True
-        cnxn_string = (
-            'DRIVER={};'
-            'PORT={};'
-            'SERVER={};'
-            'DATABASE={};'
-            'UID={};'
-            'MARS_Connection={};'
-            'PWD={}').format(
-                '{ODBC Driver 17 for SQL Server}',
-                self.port(),
-                self.server(),
-                self.database(),
-                self.credentials().userid(),
-                'yes' if self.mars() else 'no',
-                self.credentials().password())
-        self.logger.debug('Connecting: {}'.format(cnxn_string))
-        self.cnxn = pyodbc.connect(cnxn_string)
-        self.isConnecting = False
-
-    def disconnect(self):
-        if self.cnxn:
-            self.cnxn.close()
-            self.cnxn = None
-
-    def stop(self):
-        super().stop()
-        self.disconnect()
-
-    def validate_column(self, column, table, cursor):
-        """ Makes sure column belongs to table
-
-        Args:
-            column (str): column in question
-            table (str): table name
-            cursor (pyodbc.Cursor): active cursor
-
-        Returns:
-            True/False
-        """
-        columns = self._table_colums.get(table)
-        if columns is None:
-            columns = [column.column_name for column in cursor.columns(table)]
-            self._table_colums[table] = columns
-
-        if column not in columns:
-            cursor.close()
-            raise ValueError(
-                '\"{}\" is not a valid column in table \"{}\".'.format(
-                    column, table))
-
-        return column
+import pyodbc
+from nio.block.base import Block
+from nio.properties import (StringProperty, IntProperty, BoolProperty, VersionProperty, ObjectProperty, PropertyHolder)
+from nio.util.discovery import not_discoverable
+
+class Connection(PropertyHolder):
+
+    server = StringProperty(title='Server', default='[[MSSQL_SERVER]]', order=1)
+    port = IntProperty(title='Port', default='[[MSSQL_PORT]]', order=2)
+    database = StringProperty(title='Database', default='[[MSSQL_DB]]', order=3)
+    userid = StringProperty(title='User ID', allow_none=True, default='[[MSSQL_USER]]', order=4)
+    password = StringProperty(title="Password", allow_none=True, default='[[MSSQL_PWD]]', order=5)
+    mars = BoolProperty(title='Enable Multiple Active Result Sets', default=False, order=6)
+
+@not_discoverable
+class MSSQLBase(Block):
+
+    version = VersionProperty('1.0.0')

```

Comment:
@mattdodge does a `@not_discoverable` block need a version?

Reply:
Obviously I'm not Matt, but I think it does. One example of why might be that it has configurable Properties, so plausibly not all configurations are compatible with all revisions of the code.

Reply:
In this case I can find precedence: https://github.com/nio-blocks/state/blob/master/state_base.py#L27

Reply:
well, what I mean, is that you cannot create/discover an instance of this block by itself—I suppose in OO parlance, it's more or less an `abstract` class. In this particular case, the version property on this base class is going to get more or less obliterated by the inheriting classes, so I was wondering if it's even necessary.

Reply:
Oh yea that's a good point.

Reply:
It does not need a version property but if it has one the version will be inherited by subclassed blocks. Typically we won't version parent blocks though, that allows a sub block to version bump without other sub blocks doing so. In this case I would think we'd want version props only on the sub blocks.